extensions/default/src/MergeDataSource/index.js

```diff
@@ -0,0 +1,195 @@
+   * @param {Array} args - The arguments to pass to the function on each data source.
+   * @returns {Array} - An array of the results from each data source.
+   */
+  const callForAllDataSources = (path, args) => {
+    const dataSourceDefs = Object.values(extensionManager.dataSourceDefs);
+    const mergedData = [];
+    for (const dataSourceDef of dataSourceDefs) {
+      const { configuration, sourceName } = dataSourceDef;
+      if (configuration && dataSourceNames.includes(sourceName)) {
+        const [dataSource] = extensionManager.getDataSources(sourceName);
+        const func = get(dataSource, path);
+        const data = func.apply(dataSource, args);
+        mergedData.push(data);
+      }
+    }
+    return mergedData.flat();
+  };
+
+  /**
+   * Calls the default data source with the given path and arguments.
+   * @param {string} path - The path to the function to be executed.
+   * @param {Array} args - The arguments to be passed to the function.
+   * @param {string} defaultDataSourceName - The name of the default data source.
+   * @returns {*} - The result of executing the function.
+   */
+  const callForDefaultDataSource = (path, args, defaultDataSourceName) => {
+    const [dataSource] = extensionManager.getDataSources(defaultDataSourceName);
+    const func = get(dataSource, path);
+    return func.apply(dataSource, args);
+  };
+
+  /**
+   * Calls a method on the data source associated with the given RetrieveAETitle or the default data source name.
+   *
+   * @param {string} path - The path of the method to call.
+   * @param {Array} args - The arguments to pass to the method.
+   * @param {string} defaultDataSourceName - The name of the default data source.
+   * @returns {*} - The result of calling the method on the data source.
+   */
+  const callByRetrieveAETitle = (path, args, defaultDataSourceName) => {
+    const [displaySet] = args;
+    const seriesMetadata = DicomMetadataStore.getSeries(
+      displaySet.StudyInstanceUID,
+      displaySet.SeriesInstanceUID
+    );
+    const [dataSource] = extensionManager.getDataSources(
+      seriesMetadata.RetrieveAETitle || defaultDataSourceName
+    );
+    return dataSource[path](...args);
+  };
+
+  const implementation = {
+    initialize: (...args) => callForAllDataSources('initialize', args),
+    query: {
+      studies: {
+        search: (...args) => callForAllDataSourcesAsync('query.studies.search', args),
+      },
+      series: {
+        search: (...args) => callForAllDataSourcesAsync('query.series.search', args),
+      },
+      instances: {
+        search: (...args) => callForAllDataSourcesAsync('query.instances.search', args),
+      },
+    },
+    retrieve: {
+      bulkDataURI: (...args) => callForAllDataSourcesAsync('retrieve.series.metadata', args),
+      directURL: (...args) => callForAllDataSourcesAsync('retrieve.directURL', args),
+      series: {
+        metadata: (...args) => callForAllDataSourcesAsync('retrieve.series.metadata', args),
```

Comment:
The metadata will be for a single data source unless you merge multiple same series into one.

Reply:
This method checks for retrieveAETitle and filters out the other data sources, in the end is going to be one dataSource. This is necessary since we don't know which data source to call beforehand.